homeassistant/components/heos/__init__.py

```diff
@@ -219,6 +228,125 @@ def update_ids(self, mapped_ids: dict[int, int]):
+
+    def __init__(self, hass, controller):
+        """Init group manager."""
+        self._hass = hass
+        self._group_membership = {}
+        self.controller = controller
+
+    def _get_entity_id_to_player_id_map(self) -> dict:
+        """Return a dictionary which maps all HeosMediaPlayer entity_ids to player_ids."""
+        return {v: k for k, v in self._hass.data[DOMAIN][DATA_ENTITY_ID_MAP].items()}
+
+    async def async_get_group_membership(self):
+        """Return a dictionary which contains all group members for each player as entity_ids."""
+        group_info_by_entity_id = {
+            player_entity_id: []
+            for player_entity_id in self._get_entity_id_to_player_id_map()
+        }
+
+        try:
+            groups = await self.controller.get_groups(refresh=True)
+        except HeosError as err:
+            _LOGGER.error("Unable to get HEOS group info: %s", err)
+            return group_info_by_entity_id
+
+        player_id_to_entity_id_map = self._hass.data[DOMAIN][DATA_ENTITY_ID_MAP]
+        for group in groups.values():
+            leader_entity_id = player_id_to_entity_id_map.get(group.leader.player_id)
+            member_entity_ids = [
+                player_id_to_entity_id_map[member.player_id]
+                for member in group.members
+                if member.player_id in player_id_to_entity_id_map
+            ]
+            # Make sure the group leader is always the first element
+            group_info = [leader_entity_id, *member_entity_ids]
+            if leader_entity_id:
+                group_info_by_entity_id[leader_entity_id] = group_info
+                for member_entity_id in member_entity_ids:
+                    group_info_by_entity_id[member_entity_id] = group_info
+
+        return group_info_by_entity_id
+
+    async def async_join_players(
+        self, leader_entity_id: str, member_entity_ids: list[str]
+    ) -> None:
+        """Create a group with `leader_entity_id` as group leader and `member_entity_ids` as member players."""
+        entity_id_to_player_id_map = self._get_entity_id_to_player_id_map()
+        leader_id = entity_id_to_player_id_map.get(leader_entity_id)
+        member_ids = [
+            entity_id_to_player_id_map[member]
+            for member in member_entity_ids
+            if member in entity_id_to_player_id_map
+        ]
+
+        if leader_id:
+            try:
+                await self.controller.create_group(leader_id, member_ids)
+            except HeosError as err:
+                _LOGGER.error(
+                    "Failed to group %s with %s: %s",
+                    leader_entity_id,
+                    member_entity_ids,
+                    err,
+                )
+
+    async def async_unjoin_player(self, player_entity_id: str):
+        """Remove `player_entity_id` from any group."""
+        player_id = self._get_entity_id_to_player_id_map().get(player_entity_id)
+
+        if player_id:
```

Comment:
If there's no player id we should raise an exception with a message as argument. If the error is due to invalid input we raise `ValueError` other `HomeAssistantError`.

Reply:
Alright, I am now raising `HomeAssistantError`. Invalid input (invalid entity_ids) will usually not make it's way down to this method as it's filtered out by the service call `media_player.join`/`media_player.unjoin`.